cmd/crank/beta/validate/validate.go

```diff
@@ -17,102 +17,132 @@ limitations under the License.
+			if err != nil {
+				return nil, nil, err
 			}
+
+			structurals[gvk] = structural
 		}
 	}
 
-	return validators, nil
+	return validators, structurals, nil
 }
 
 // SchemaValidation validates the resources against the given CRDs
 func SchemaValidation(resources []*unstructured.Unstructured, crds []*extv1.CustomResourceDefinition, skipSuccessLogs bool) error {
-	schemaValidators, err := newValidators(crds)
+	schemaValidators, structurals, err := newValidatorsAndStructurals(crds)
 	if err != nil {
 		return errors.Wrap(err, "cannot create schema validators")
 	}
 
-	failure, warning := 0, 0
+	failure, missingSchemas := 0, 0
 
 	for i, r := range resources {
-		resourceValidators, ok := schemaValidators[r.GetObjectKind().GroupVersionKind()]
+		gvk := r.GetObjectKind().GroupVersionKind()
+		sv, ok := schemaValidators[gvk]
 		if !ok {
-			warning++
+			missingSchemas++
 			fmt.Println("[!] could not find CRD/XRD for: " + r.GroupVersionKind().String())
 			continue
 		}
 
 		rf := 0
-		for _, v := range resourceValidators {
-			re := v.Validate(&resources[i])
-			for _, e := range re.Errors {
+
+		for _, v := range sv {
+			re := validation.ValidateCustomResource(nil, r, *v)
+			for _, e := range re {
 				rf++
-				fmt.Printf("[x] validation error %s, %s : %s\n", r.GroupVersionKind().String(), r.GetAnnotations()[composite.AnnotationKeyCompositionResourceName], e.Error())
+				fmt.Printf("[x] schema validation error %s, %s : %s\n", r.GroupVersionKind().String(), getResourceName(r), e.Error())
 			}
-		}
 
-		if rf == 0 && !skipSuccessLogs {
-			fmt.Printf("[✓] %s, %s validated successfully\n", r.GroupVersionKind().String(), r.GetAnnotations()[composite.AnnotationKeyCompositionResourceName])
-		} else {
-			failure++
+			s := structurals[gvk] // if we have a schema validator, we should also have a structural
+
+			celValidator := cel.NewValidator(s, true, celconfig.PerCallLimit)
+			re, _ = celValidator.Validate(context.TODO(), nil, s, resources[i].Object, nil, celconfig.PerCallLimit)
+			for _, e := range re {
+				rf++
+				fmt.Printf("[x] CEL validation error %s, %s : %s\n", r.GroupVersionKind().String(), getResourceName(r), e.Error())
+			}
+
+			if rf == 0 && !skipSuccessLogs {
+				fmt.Printf("[✓] %s, %s validated successfully\n", r.GroupVersionKind().String(), getResourceName(r))
+			} else {
+				failure++
+			}
 		}
 	}
 
-	fmt.Printf("%d error, %d warning, %d success cases\n", failure, warning, len(resources)-failure-warning)
+	fmt.Printf("Total %d resources: %d missing schemas, %d success cases, %d failure cases\n", len(resources), missingSchemas, len(resources)-failure-missingSchemas, failure)
 
 	if failure > 0 {
 		return errors.New("could not validate all resources")
 	}
 
 	return nil
 }
+
+func getResourceName(r *unstructured.Unstructured) string {
+	if r.GetName() != "" {
+		return r.GetName()
+	}
+
+	// fallback to composition resource name
+	return r.GetAnnotations()[composite.AnnotationKeyCompositionResourceName]
+}
```

Comment:
what do you think about the following logic instead:
- no name, but annotation => `"" (annotation)`
- name, no annotation => `"name"`
- name and annotation => `"name" (annotation)`

both name and annotation could be useful to function users as they can now set the name of resources too.

Reply:
I prefer keeping it simple and not making the output hard to read. Since there can be only one resource with the same name and GVK, users can check the annotation if they want.

Reply:
I don't think it would be hard to read and in a scenario where I'm piping the render output into this there would be no manifests to check, I'd have to rerun the command without piping to see both.

At least I'd quote (`%q`) the name so that if both the annotation and the name are missing you don get a dangling comma and it's easier to see the name is missing: `iam.aws.upbound.io/v1beta1, Kind=User,  validated successfully` => `iam.aws.upbound.io/v1beta1, Kind=User, "" validated successfully`.